Opt-out updated bootimage for GCP and AWS (#93065)
openshift/openshift-docs: modules/mco-update-boot-images-disable.adoc

// Module included in the following assemblies:
//
// * machine-configuration/mco-update-boot-images.adoc
// * nodes/nodes-nodes-managing.adoc

:_mod-docs-content-type: PROCEDURE
[id="mco-update-boot-images-disable_{context}"]
= Disabling updated boot images

> By default, for {gcp-first} and {aws-first} clusters, by default the Machine Config Operator (MCO) manages and updates the boot image in the machine sets in your cluster. This means that the MCO updates the boot images whenever you update your cluster.

Reviewer: same issue as https://github.com/openshift/openshift-docs/pull/93065/files#r2091656145

You can disable this default behavior on some or all of your machine sets by editing the `MachineConfiguration` object. When disabled, the Machine Config Operator (MCO) no longer manages the boot image in your cluster and no longer updates the boot image with each cluster update.

> Disabling this feature does not rollback the nodes or machine sets to the originally-installed boot image. The machine sets retain the boot image version that was present when the feature was disabled and is not updated if the cluster is upgraded to a new {product-title} version in the future. Existing nodes are unchanged, retaining their current boot image.

Reviewer: ```suggestion
Disabling this feature does not rollback the nodes or machine sets to the originally-installed boot image. The machine sets retain the boot image version that was present when the feature was disabled and is not updated if the cluster is upgraded to a new {product-title} version in the future. As a boot image only use when a node is scaled up, this feature has no effect on existing nodes.
```

> If you disable this feature after some nodes have been created with the new boot image version, any existing nodes retain their current boot image. Disabling this feature does not rollback the nodes or machine sets to the originally-installed boot image. The machine sets retain the boot image version that was present when the feature was enabled and is not updated again when the cluster is upgraded to a new {product-title} version in the future. 

Reviewer: Just want to mention the same concept as https://github.com/openshift/openshift-docs/pull/93065/files#r2091637012; i.e. onces nodes are born, their boot image cannot change. 
```suggestion
If you disable this feature after some nodes have been created with the new boot image, the machinesets will retain their current boot image. Disabling this feature does not rollback the machine sets to the originally-installed boot image. The machine sets retain the boot image version that was present when the feature was enabled and is not updated again when the cluster is upgraded to a new {product-title} version in the future. 
```

Author: > If you disable this feature after some nodes have been created with the new boot image, the machinesets will retain their current boot image.

The machine sets will retain their current boot image after disabling the feature, whether new nodes are created or not, right? 
This was another mention that the existing nodes are unaffected by the opt-out.

Reviewer: > The machine sets will retain their current boot image after disabling the feature, whether new nodes are created or not, right?
Correct!

> This was another mention that the existing nodes are unaffected by the opt-out.
Yeah, this was me emphasizing that existing nodes are entirely unaffected by this feature.

  managedBootImages: <1>
    machineManagers: []
----
<1> Remove the parameters that are listed under `machineManagers` and add the `[]` characters to disable the updating of boot images.

* Optional: Disable the default behavior for specific machine sets:
+
[source,yaml]
----
apiVersion: operator.openshift.io/v1
kind: MachineConfiguration
metadata:
  name: cluster
  namespace: openshift-machine-config-operator
spec:
# ...
  managedBootImages: <1>
    machineManagers:
    - apiGroup: machine.openshift.io <2>
      resource: machinesets <3>
      selection:
        mode: Partial <4>
        partial:
          machineResourceSelector: <5>
            matchLabels:
              region: "east"
----
<1> Configures the boot image update feature.
<2> Specifies an API group. This must be `machine.openshift.io`.
<3> Specifies the resource within the specified API group to apply the change. This must be `machinesets`.
<4> Specifies that the default behavior is disabled for specific machine sets.

> <5> Specifies that any machine set with this label is to be updated. A positive match of the label selector disables the default behavior for that machine set.

Reviewer: I'm not sure what you mean by "default behavior" here. With this configuration, any machineset without the label will not be updated. We don't expect many users to disable boot image updates in such a manner though, so I think we should probably drop this section too. Did you mean negative match here?

Author: @djoshy 
> we should probably drop this section too

Which section? Remove the _Disable the default behavior for specific machine sets_ step and just go with _Disable the behavior for all machine sets_?

Reviewer: It seems like this section was removed from the re-enable section instead?

Author: > Did you mean negative match here?

Oh wait, if the machine set has the label, the default behavior would be enabled, right. So, to "Disable the default behavior for specific machine sets", add matchLabels that are not present in those machine sets?

Reviewer: Correct, we expect users to do this if they want to "canary" test their machinesets. So, I think this subsection would be a better fit for "re-enabling" rather than "disabling". 

Over time though, I hope most customers try to use the "All" mode, once they're happy with the feature :smile:
